File: Javascript/22createLinkedList.js


// iterative
// n = length of values
// time:  O(n)
// space: O(n)
const createLinkedList = (values) => {
  const dummyHead = new Node(null);
  let tail = dummyHead;
  for (let val of values) {
    tail.next = new Node(val);
    tail = tail.next;
  }
  return dummyHead.next;
};

// --------------------------------------- // 

// recursive
// n = length of values
// time:  O(n)
// space: O(n)
const createLinkedList = (values, i = 0) => {
  if (i === values.length) return null;
  const head = new Node(values[i]);
  head.next = createLinkedList(values, i + 1);
  return head;
};

File: Python/22create_linked_list.py
class Node:
  def __init__(self, val):
    self.val = val
    self.next = None


# recursive
# n = length of values
# Time: O(n)
# Space: O(n)
def create_linked_list(values, i= 0):
  if i == len(values):
    return None
  value = values[i]
  new_head = Node(value)
  new_head.next = create_linked_list(values, i+1)
  return new_head


# iterative
# n = length of values
# Time: O(n)
# Space: O(n)
def create_linked_list(values):
  dummy = Node(0)
  current = dummy
  for value in values:
    current.next = Node(value)
    current = current.next

  return dummy.next
